### PR TITLE
add a file tree generator

### PR DESCRIPTION
**Pull request by safchain**
Signed-off-by: Sylvain Afchain <sylvain.afchain@datadoghq.com>

### What does this PR do?

Add a file tree generator. First the generator generates a tree, then it randomly generates file access and folder rename.

### Motivation

What inspired you to submit this pull request?

### Related issues

A list of issues either fixed, containing architectural discussions, otherwise relevant
for this Pull Request.

### Additional Notes

Anything else we should know when reviewing?

### CODE REVIEW

File: src/generator/file_tree.rs [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+    }
+    Ok(())
+}
+
+#[inline]
+fn rnd_node_name(rng: &mut StdRng, dir: &Path, len: usize) -> PathBuf {
+    let name = Alphanumeric.sample_string(rng, len);
+    let mut node = dir.to_path_buf();
+    node.push(name);
+    node
```
Comment (paulcacheux) [inline src/generator/file_tree.rs:225]:
```suggestion
    let name = Alphanumeric.sample_string(rng, len);
    dir.join(name)
```

**Review by GeorgeHahn — CHANGES_REQUESTED** [review]
This is looking great! There's one change I'd like to ask for. We try to do as much work as possible at startup to keep generators fast at runtime. Could you switch the node name generation to create a large list at startup and pull from that as needed?

File: src/generator/file_tree.rs (hunk 1 above)
Comment (safchain) [reply src/generator/file_tree.rs]: Done, thanks

**Comment by safchain** [comment]
@GeorgeHahn thanks for having a look. I submitted a new version, the list of nodes (file/folder) are now generated in the new function.

File: src/generator/file_tree.rs [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    /// The maximum number of nodes (folder/file) created
+    pub max_nodes: NonZeroU32,
+    /// The name length of the nodes (folder/file) without the extension
+    pub name_len: NonZeroUsize,
+    /// The root folder where the nodes will be created
+    pub root: String,
+    /// The number of nodes opened per second
+    pub open_per_second: NonZeroU32,
+    /// The number of rename per second
+    pub rename_per_second: NonZeroU32,
```
Comment (blt) [inline src/generator/file_tree.rs:63]: I wouldn't be mad at some reasonable defaults for these non-zero values, but that we can tidy up in a follow-up.

Comment (safchain) [reply src/generator/file_tree.rs]: I added some default

File: src/generator/file_tree.rs [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+
+                            if is_file(&node) {
+                                files.push(node);
+                            } else {
+                                folders.push(node);
+                            }
+                        },
+                        _ => {
+                            // randomly generate an access
+                            if let Some(node) = files.choose(&mut self.rng) {
```
[the remark this answers is not in the document]
Comment (safchain) [reply src/generator/file_tree.rs]: I move the file/folder allocations to the `new`  function

**Review by blt — APPROVED** [review]
This is looking great to me. I'll be happy to see it merged.

**Review by GeorgeHahn — APPROVED** [review]
This looks fantastic! Thank you @safchain.

**Pull request merged**